src/sage/combinat/partition.py

```diff
@@ -6288,6 +6295,17 @@ def __classcall_private__(cls, n=None, **kwargs):
                 elif 'max_part' in kwargs and 'max_length' in kwargs:
                     return PartitionsInBox(kwargs['max_length'], kwargs['max_part'])
 
+            # IntegerListsLex does not deal well with infinite sets,
+            # so we use a class inheriting from Partitions, and
+            # normalize arguments here
+            if 'outer' in kwargs:
+                kwargs['outer'] = tuple([infinity if x is infinity else ZZ(x)
+                                         for x in kwargs['outer']])
+            if 'inner' in kwargs:
+                kwargs['inner'] = tuple([ZZ(x) for x in kwargs['inner']])
+                if min(kwargs['inner']) < 1:
+                    raise ValueError("the inner shape must contain only positive integers")
```

Comment:
```suggestion
                if min(kwargs['inner']) < 0:
                    raise ValueError("the inner shape must contain only nonnegative integers")
```
See above. Although I am not sure I find this to be a worthwhile restriction. Everything still makes sense, and the input does not need to be a partition (or at least it shouldn't need to be).

Reply:
I don't mind - the main reason for the preprocessing is that `Partition` inherits from  `UniqueRepresentation`.
However, I think that implicitly `inner` must be a partition: specifying `inner=[3, 3, 3, 1, 2, 2]` is the same as specifying `inner=[3, 3, 3, 2, 2, 2]`.  Similarly, I think that `outer` should also be a partition, except that the first few parts may be infinity.

I would expect that any non-partition input would be an unconcious decision and possibly hide a bug.

Reply:
I made the change, but now we have another problem: per documentation on line 154:
```
Finally, here are the partitions of `4` with ``[1,1,1]`` as an inner
bound (i. e., the partitions of `4` containing the partition ``[1,1,1]``).
Note that ``inner`` sets ``min_length`` to the length of its argument::

    sage: Partitions(4, inner=[1,1,1]).list()
    [[2, 1, 1], [1, 1, 1, 1]]
```

However, doctest on line 6052:
```
        sage: Partitions(5, inner=[2,0,0,0,0,0]).list()
        [[5], [4, 1], [3, 2], [3, 1, 1], [2, 2, 1], [2, 1, 1, 1]]
```

Reply:
> I don't mind - the main reason for the preprocessing is that `Partition` inherits from `UniqueRepresentation`. However, I think that implicitly `inner` must be a partition: specifying `inner=[3, 3, 3, 1, 2, 2]` is the same as specifying `inner=[3, 3, 3, 2, 2, 2]`. Similarly, I think that `outer` should also be a partition, except that the first few parts may be infinity.
> 
> I would expect that any non-partition input would be an unconcious decision and possibly hide a bug.

While normalizing the input, then I think converting to a partition would be good. I think this is safe if we document the behavior (or at least a user should not be surprised in this case).

Reply:
> I made the change, but now we have another problem: per documentation on line 154:
> 
> ```
> Finally, here are the partitions of `4` with ``[1,1,1]`` as an inner
> bound (i. e., the partitions of `4` containing the partition ``[1,1,1]``).
> Note that ``inner`` sets ``min_length`` to the length of its argument::

If I interpret "length" as meaning the length of the partition, then there is no problem. It isn't exactly clear if it is referring to that or `len()`. I think it is better (mathematically and programmatically) to treat it as the length of the partition.